compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -602,4 +612,61 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         }
         self.flounder(&responses)
     }
+
+    #[instrument(level = "debug", skip(self), ret)]
+    pub(super) fn evaluate_alias_bound_self_is_well_formed<G: GoalKind<'tcx>>(
+        &mut self,
+        goal: Goal<'tcx, G>,
+    ) -> QueryResult<'tcx> {
+        match *goal.predicate.self_ty().kind() {
+            ty::Alias(ty::Projection, projection_ty) => {
+                let mut param_env_candidates = vec![];
+                let projection_trait_ref = projection_ty.trait_ref(self.tcx());
+
+                if projection_trait_ref.self_ty().is_ty_var() {
+                    return self
+                        .evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS);
+                }
+
+                let trait_goal = goal.with(
+                    self.tcx(),
+                    ty::TraitPredicate {
+                        trait_ref: projection_trait_ref,
+                        constness: ty::BoundConstness::NotConst,
+                        polarity: ty::ImplPolarity::Positive,
+                    },
+                );
+
+                // FIXME: We probably need some sort of recursion depth incr here.
+                // Can't come up with an example yet, though, and the worst case
+                // we can have is a compiler stack overflow...
+                self.assemble_param_env_candidates(trait_goal, &mut param_env_candidates);
+                self.assemble_alias_bound_candidates(trait_goal, &mut param_env_candidates);
```

Comment:
We need to consider nested alias bounds here I think - https://hackmd.io/frHMYxD5S4-rWlxkx_eWNw#3-Nested-alias-bound-candidates